viewer/packages/camera-manager/src/Flexible/FlexibleCameraManager.ts

```diff
@@ -0,0 +1,551 @@
+ */
+
+import { Box3, PerspectiveCamera, Quaternion, Raycaster, Vector2, Vector3, Scene, Ray } from 'three';
+
+import { FlexibleControls } from './FlexibleControls';
+import { FlexibleControlsOptions } from './FlexibleControlsOptions';
+
+import {
+  assertNever,
+  EventTrigger,
+  InputHandler,
+  disposeOfAllEventListeners,
+  PointerEventData,
+  fitCameraToBoundingBox,
+  clickOrTouchEventOffset
+} from '@reveal/utilities';
+
+import { getNormalizedPixelCoordinates } from '@reveal/utilities';
+import {
+  CameraChangeDelegate,
+  CameraEventDelegate,
+  CameraManagerCallbackData,
+  CameraManagerEventType,
+  CameraState,
+  CameraStopDelegate
+} from './../types';
+import { CameraManagerHelper } from './../CameraManagerHelper';
+import { CameraManager } from './../CameraManager';
+import { FlexibleControlsType } from './FlexibleControlsType';
+import { FlexibleMouseActionType } from './FlexibleMouseActionType';
+import { DebouncedCameraStopEventTrigger } from '../utils/DebouncedCameraStopEventTrigger';
+import { FlexibleCameraMarkers } from './FlexibleCameraMarkers';
+import { moveCameraTargetTo, moveCameraTo } from './moveCamera';
+import { IFlexibleCameraManager } from './IFlexibleCameraManager';
+
+type RaycastCallback = (x: number, y: number, pickBoundingBox: boolean) => Promise<CameraManagerCallbackData>;
+
+/**
+ * Flexible implementation of {@link CameraManager}. The user can switch between Orbit, FirstPersion or OrbitInCenter
+ * Supports automatic update of camera near and far planes and animated change of camera position and target.
+ * @beta
+ */
+
+export class FlexibleCameraManager implements IFlexibleCameraManager {
+  //================================================
+  // INSTANCE FIELDS:
+  //================================================
+
+  private readonly _events = { cameraChange: new EventTrigger<CameraChangeDelegate>() };
+  private readonly _stopEventTrigger: DebouncedCameraStopEventTrigger;
+  private readonly _controls: FlexibleControls;
+  private readonly _camera: PerspectiveCamera;
+  private readonly _domElement: HTMLElement;
+  private readonly _inputHandler: InputHandler;
+  private readonly _markers?: undefined | FlexibleCameraMarkers;
+  private readonly _currentBoundingBox: Box3 = new Box3();
+  private _isDisposed = false;
+  private _nearAndFarNeedsUpdate = false;
+  private _isEnabled = true;
+
+  // For the wheel event
+  private _prevTime = 0;
+  private readonly _prevCoords = new Vector2();
+
+  //================================================
+  // INSTANCE FIELDS: Events
+  //================================================
+
+  private readonly _raycastCallback: RaycastCallback;
+
+  //================================================
+  // CONSTRUCTOR
+  //================================================
+
+  constructor(
+    domElement: HTMLElement,
+    inputHandler: InputHandler,
+    raycastCallback: RaycastCallback,
+    camera?: PerspectiveCamera,
+    scene?: Scene
```

Comment:
Is there a reason we want `scene` to be optional?

Reply:
It seems a little unnecessary - I don't see how we would use the FlexibleCameraManager if there is no scene (it depends on the bounds etc. of the scene too). But whatever